### PR TITLE
feat: Image Annotator

### LINKED ISSUE #3362: [Feature] Image Annotator
## Summary

Users need a widget to annotate images

### PR DESCRIPTION
**Pull request by wmdev0808**
## Description
A widget that can be used for labelling objects inside images. 
Features:
- Ability to draw inside image using a polygon
- Store coordinates of the box and send it an API request

Fixes #3362 

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## How Has This Been Tested?

- Test A
- Test B

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: feature/3362-image-annotator 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 55.14 **(-0.01)** | 37.1 **(0.01)** | 33.98 **(0.01)** | 55.66 **(0.01)**
 :red_circle: | app/client/src/utils/autocomplete/EntityDefinitions.ts | 26.32 **(-0.71)** | 0 **(0)** | 12.5 **(-1.79)** | 25.71 **(-0.76)**
 :sparkles: :new: | **app/client/src/widgets/ImageAnnotatorWidget/constants.ts** | **100** | **100** | **100** | **100**
 :sparkles: :new: | **app/client/src/widgets/ImageAnnotatorWidget/index.ts** | **100** | **100** | **100** | **100**
 :sparkles: :new: | **app/client/src/widgets/ImageAnnotatorWidget/component/index.tsx** | **38.46** | **50** | **0** | **38.46**
 :sparkles: :new: | **app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx** | **40** | **0** | **41.67** | **43.18**</details>

### CODE REVIEW

**Comment by somangshu** [comment]
@wmdev0808 jest tests failing because of related changes please check and fix

**Review by somangshu — APPROVED** [review]
Approving to run the tests. @SatishGandham will review the code

**Review by SatishGandham — APPROVED** [review]

**Comment by somangshu** [comment]
@wmdev0808 can you clear the merge conflicts

@vivekverma2312 @YogeshJayaseelan you guys need to prioritise tasks properly here, this PR got closed due to inactivity. We now have to go through the whole flow. the icon will need to change, There might me new reviews required. Lets catchup to strategise this

cc @yatinappsmith

**Comment by wmdev0808** [comment]
https://github.com/appsmithorg/appsmith/actions/runs/1416920477

**Review by somangshu — COMMENTED** [review]
Good work @wmdev0808,
Feedback on the deploy preview:
- The onAnnotationChange is called every-time the mouse hovers on the widget, this is not desired, If we cant make this appear only when the annotation changes, Which is probably not offered as a functionality, we should remove this
- Provide all selectors available by default in the package: `[RectangleSelector, PointSelector, OvalSelector]`, Expose the API to user for choosing one
- There is a property available annotations & activeAnnotations, if this lets the dev place annotation programmatically, we should expose this API (use case: re-enforcement learning) 
- Provide disableAnnotation API for the dev to disable creating annotations
- @momcilo-appsmith @UXFalcon we can change the way the input looks, Please take a look an let us know if that is required
- expose `alt` API for the user


@SatishGandham kindly review the code, I also have the following question, Let me know what are your thoughts
- Is it wise to connect selector logic to redux? https://github.com/RobinNagpal/react-image-annotation-ts#connecting-selector-logic-to-reduxmobx
- We should probably fork the repo, Since is is not a very popular package

File: app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+            label: "Image URL",
+            controlType: "INPUT_TEXT",
+            placeholderText: "Enter URL / Base64",
+            isBindProperty: true,
+            isTriggerProperty: false,
+            validation: { type: ValidationTypes.IMAGE_URL },
+          },
+          {
+            propertyName: "isDisabled",
+            helpText: "Disables input to the widget",
```
Comment (SatishGandham) [inline app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx:30]: Do you mean disable annotations?

File: app/client/src/widgets/ImageAnnotatorWidget/component/index.tsx [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    onChange,
+    onReset,
+    onSubmit,
+  } = props;
+
+  useEffect(() => {
+    onReset();
+  }, [imageUrl]);
+
+  const handleChange = useCallback(
```
Comment (SatishGandham) [inline app/client/src/widgets/ImageAnnotatorWidget/component/index.tsx:43]: We are not doing any additional work on handleChange and submit, why not pass these props directly to Annotation?

File: app/client/src/widgets/ImageAnnotatorWidget/component/index.tsx [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  const handleSubmit = useCallback(
+    (annotation: IAnnotation) => {
+      onSubmit(annotation);
+    },
+    [onSubmit],
+  );
+
+  return (
+    <ImageAnnotatorContainer>
+      {disabled && <DisabledOverlay />}
```
Comment (SatishGandham) [inline app/client/src/widgets/ImageAnnotatorWidget/component/index.tsx:59]: What does disabled mean?

**Review by SatishGandham — CHANGES_REQUESTED** [review]

File: app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx (hunk 1 above)
Comment (wmdev0808) [reply app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx:30]: No, it is just for disabling widget as we did with other widgets

File: app/client/src/widgets/ImageAnnotatorWidget/component/index.tsx (hunk 2 above)
Comment (wmdev0808) [reply app/client/src/widgets/ImageAnnotatorWidget/component/index.tsx:43]: It is just for avoid unnecessary re-render

File: app/client/src/widgets/ImageAnnotatorWidget/component/index.tsx (hunk 3 above)
Comment (wmdev0808) [reply app/client/src/widgets/ImageAnnotatorWidget/component/index.tsx:59]: It means if user disabled the widget from the property pane

**Comment by SatishGandham** [comment]
@somangshu @wmdev0808 ,

- How does the user show these saved annotations?
- _Provide all selectors available by default in the package: [RectangleSelector, PointSelector, OvalSelector], Expose the API to user for choosing one_. To which user, Appsmith dev or end user? I think we will need designs to expose to end user.
- Connecting to the store is not required, state can be maintained in the component itself.

**Comment by somangshu** [comment]
> @somangshu @wmdev0808 ,
> 
> * How does the user show these saved annotations? @wmdev0808 _**does activeAnnotation / Annotation API allow us to do the same**_
> * _Provide all selectors available by default in the package: [RectangleSelector, PointSelector, OvalSelector], Expose the API to user for choosing one_. To which user, Appsmith dev or end user? I think we will need designs to expose to end user. @SatishGandham _**this is only for the dev right now**_
> * Connecting to the store is not required, state can be maintained in the component itself. 👍

**Comment by wmdev0808** [comment]
https://github.com/appsmithorg/appsmith/actions/runs/1444698732

**Comment by somangshu** [comment]
Great value add with the default annotation feature. This will be helpful for the devs and makes the feature more complete. My feedback around it is:
- The property should be below the `alt text` property
- I was thinking about having the actual object separated like how we are trying to doing in the chart widget, But this is not a good idea because the input / output of the annotation object is same here. We need to do a real good job in create a documentation around this. @wmdev0808 can you draft the initial version of that. @Pranay105 we will need your help to polish this further. 

**_Issue_**:
- the widget is not working on the view mode. seeing a console error
- once we create an annotation We can see the widget grouping boundary enabled until the canvas is clicked again

https://user-images.githubusercontent.com/11089579/141248801-0f8dd708-c9cc-4ef0-8640-d55a5ed55aaa.mov

**Review by somangshu — DISMISSED** [review]
Left comment

File: app/client/src/widgets/ImageAnnotatorWidget/component/index.tsx (hunk 2 above)
Comment (SatishGandham) [reply app/client/src/widgets/ImageAnnotatorWidget/component/index.tsx:53]: It doesn't solve the re-render issue.
It has to re-render anyway if onChange/onSubmit change.

**Comment by wmdev0808** [comment]
https://github.com/appsmithorg/appsmith/actions/runs/1448986612

File: app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx [hunk 4]
```diff
@@ -217,6 +187,26 @@ class ImageAnnotatorWidget extends BaseWidget<
...
+            label: "Visible",
+            controlType: "SWITCH",
+            isJSConvertible: true,
+            isBindProperty: true,
+            isTriggerProperty: false,
+            validation: { type: ValidationTypes.BOOLEAN },
+          },
+          {
+            propertyName: "isAnnotationDisabled",
+            helpText: "Set to true to disable creating of annotations",
```
Comment (somangshu) [inline app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx:202]: Change to `Controls if user is allowed to draw annotations`

File: app/client/src/widgets/ImageAnnotatorWidget/icon.svg [hunk 5]
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" width="32" height="32" fill="none" viewBox="0 0 32 32"><path fill="#F2FAFF" fill-opacity=".9" d="M28 25.3333V6.66667C28 5.2 26.8 4 25.3333 4H6.66667C5.2 4 4 5.2 4 6.66667V25.3333C4 26.8 5.2 28 6.66667 28H25.3333C26.8 28 28 26.8 28 25.3333ZM11.3333 18L14.6667 22.0133L19.3333 16L25.3333 24H6.66667L11.3333 18Z"/></svg>
```
Comment (somangshu) [inline app/client/src/widgets/ImageAnnotatorWidget/icon.svg:1]: @momcilo-appsmith @UXFalcon need help with the new icon

**Comment by somangshu** [comment]
@SatishGandham @wmdev0808 this is not covered by test, Do you think anything is required?

@wmdev0808 have we created documentation for this?

**Comment by wmdev0808** [comment]
> @SatishGandham @wmdev0808 this is not covered by test, Do you think anything is required?
> 
> @wmdev0808 have we created documentation for this?

@somangshu, I didn't care about the test at the moment.
I've already created the basic documentation of the widget

File: app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx (hunk 4 above)
Comment (SatishGandham) [reply app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx:202]: @somangshu should we call this property `isAnnotationDisabled`? Isn't `isDisabled` better and consistent with other widgets?

cc: @wmdev0808

**Review by SatishGandham — CHANGES_REQUESTED** [review]
@wmdev0808  Looks good apart from the disabled property name.
Please add some test cases.

**Comment by Pranay105** [comment]
Draft an initial version of the doc. Just make sure it covers everything - functionality, behavior, exceptions, etc. I'll work on that. @somangshu @wmdev0808

File: app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx (hunk 4 above)
Comment (somangshu) [reply app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx:202]: totally agree to that, @wmdev0808 we will need to make this chagne

**Comment by somangshu** [comment]
> Draft an initial version of the doc. Just make sure it covers everything - functionality, behavior, exceptions, etc. I'll work on that. @somangshu @wmdev0808

@Pranay105 https://github.com/appsmithorg/appsmith-docs/pull/294

File: app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+            label: "Image URL",
+            controlType: "INPUT_TEXT",
+            placeholderText: "Enter URL / Base64",
+            isBindProperty: true,
+            isTriggerProperty: false,
+            validation: { type: ValidationTypes.IMAGE_URL },
+          },
+          {
+            propertyName: "imageAltText",
+            helpText: "Sets image alt attribute",
```
Comment (SatishGandham) [inline app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx:37]: @vihar , Is the language correct?

File: app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+            label: "Visible",
+            controlType: "SWITCH",
+            isJSConvertible: true,
+            isBindProperty: true,
+            isTriggerProperty: false,
+            validation: { type: ValidationTypes.BOOLEAN },
+          },
+          {
+            propertyName: "isDisabled",
+            helpText: "Sets to true to disable creating of annotations",
```
Comment (SatishGandham) [inline app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx:202]: @vihar , The language doesn't seem right. Please check.

**Comment by SatishGandham** [comment]
@Pranay105 Can you go through the property descriptions in this file once and suggest the changes
https://github.com/appsmithorg/appsmith/pull/7234/files/e98182fe7485778323753bc6633bdc559229a77c#diff-9baf5fca6574099827df5155a9dcd05166eae3d459c589936685c5f5836ce626

File: app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+  ImageAnnotatorWidgetProps,
+  WidgetState
+> {
+  static getPropertyPaneConfig() {
+    return [
+      {
+        sectionName: "General",
+        children: [
+          {
+            helpText: "Url of the image to be annotated",
```
Comment (Pranay105) [inline app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx:26]: To add an image to the canvas, enter its URL here.

File: app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx (hunk 6 above)
Comment (Pranay105) [inline app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx:37]: Enter the alternative text for the image here, giving a short description of it.

File: app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+            validation: {
+              type: ValidationTypes.TEXT,
+              params: {
+                default: "",
+              },
+            },
+          },
+          {
+            propertyName: "defaultAnnotations",
+            helpText: "Array of annotations",
```
Comment (Pranay105) [inline app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx:52]: I'm not able to understand what this property is. Can someone please explain this?

File: app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+                  },
+                },
+              },
+            },
+            evaluationSubstitutionType:
+              EvaluationSubstitutionType.SMART_SUBSTITUTE,
+          },
+          {
+            propertyName: "isVisible",
+            helpText: "Controls the visibility of the widget",
```
Comment (Pranay105) [inline app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx:192]: Controls widget's visibility on the page. When turned off, the widget will not be visible when the app is published.

File: app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx (hunk 7 above)
Comment (Pranay105) [reply app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx:202]: It enables or disables the annotation. If turned on, the user cannot annotate the image.

File: app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+            label: "Disabled",
+            controlType: "SWITCH",
+            isJSConvertible: true,
+            isBindProperty: true,
+            isTriggerProperty: false,
+            validation: { type: ValidationTypes.BOOLEAN },
+          },
+          {
+            propertyName: "selector",
+            helpText: "Sets the selector of the widget",
```
Comment (Pranay105) [inline app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx:212]: It allows you to choose the type of selector for the annotation.

File: app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx (hunk 9 above)
Comment (wmdev0808) [reply app/client/src/widgets/ImageAnnotatorWidget/widget/index.tsx:52]: @Pranay105 ,
This property is for setting default annotations array which will be used as underlying data for annotations in the first place

**Review by SatishGandham — APPROVED** [review]

**Comment by somangshu** [comment]
@wmdev0808 related tests have failed. Please check!

**Review by somangshu — APPROVED** [review]

**Pull request closed without merging**